File: lib/pitch.ts
export const midiToFrequency = (midi: number) => 440 * 2 ** ((midi - 69) / 12);
export const frequencyToMidi = (frequency: number) =>
  69 + 12 * Math.log2(frequency / 440);
export function noteName(midi: number) {
  const names = [
    'C',
    'C♯',
    'D',
    'D♯',
    'E',
    'F',
    'F♯',
    'G',
    'G♯',
    'A',
    'A♯',
    'B',
  ];
  const n = Math.round(midi);
  return names[((n % 12) + 12) % 12] + (Math.floor(n / 12) - 1);
}
/** Lowest fundamental treated as a voice: just under A1. Below is rumble, not singing. */
export const MIN_HZ = 54;
/** Highest fundamental treated as a voice, comfortably above any sung note. */
export const MAX_HZ = 2200;
/**
 * How periodic a window has to be before it counts as a note.
 *
 * Textbook YIN uses 0.10-0.15, which assumes a clean signal. A breathy voice is
 * not clean: the aperiodic part of the sound pushes the normalised difference
 * well above that even when the pitch is perfectly clear to a listener, and the
 * breathiest notes are usually the lowest ones. Measured against noise, hiss,
 * breath, mains hum and rumble, 0.30 still produced no false positives.
 */
const THRESHOLD = 0.3;

/**
 * Flatten a slow amplitude envelope while leaving the pitch periodicity alone.
 *
 * A lip trill is a normally-pitched tone whose loudness the lips flutter at
 * roughly 20-35 Hz. That flutter is a real periodicity, and a strong one: at
 * low notes it beats the vocal period, so YIN locks onto the flutter and
 * reports a note two octaves below what is being sung, or nothing at all.
 * Dividing by a short running magnitude equalises the loudness across the
 * window so only the vocal period is left for YIN to find. Vibrato, which is
 * shallower and slower, passes through untouched.
 */
function flattenEnvelope(buffer: Float32Array, sampleRate: number) {
  // About 6ms. This has to sit in a narrow band: long enough to cover most of
  // one pitch period at the bottom of the range (a 100Hz note is 10ms), or the
  // division reshapes the waveform itself and a breathy low note stops being
  // findable; short enough to still track a 20-35Hz flutter. Measured at
  // sampleRate/500 it destroyed low notes; at sampleRate/160 both survive.
  const window = Math.max(8, Math.floor(sampleRate / 160));
  const magnitude = new Float32Array(buffer.length);
  let sum = 0;
  for (let i = 0; i < buffer.length; i++) {
    sum += Math.abs(buffer[i]);
    if (i >= window) sum -= Math.abs(buffer[i - window]);
    magnitude[i] = sum / Math.min(i + 1, window);
  }
  let peak = 0;
  for (const m of magnitude) if (m > peak) peak = m;
  // Never divide by near-silence, which would only amplify noise.
  const floor = peak * 0.15;
  const out = new Float32Array(buffer.length);
  for (let i = 0; i < buffer.length; i++)
    out[i] = buffer[i] / Math.max(magnitude[i], floor);
  return out;
}

// YIN with cumulative mean normalization and parabolic interpolation.
export function detectPitch(
  buffer: Float32Array,
  sampleRate: number,
): { frequency: number; confidence: number } | null {
  if (!Number.isFinite(sampleRate) || sampleRate <= 0 || buffer.length < 1024)
    return null;
  // Reduce the lag search cost at high device sample rates. Average each block
  // before decimation; the supported singing fundamentals remain below Nyquist.
  const stride = Math.max(1, Math.floor(sampleRate / 48000));
  if (stride > 1) {
    const reduced = new Float32Array(Math.floor(buffer.length / stride));
    for (let i = 0; i < reduced.length; i++) {
      let sum = 0;
      for (let j = 0; j < stride; j++) sum += buffer[i * stride + j];
      reduced[i] = sum / stride;
    }
    buffer = reduced;
    sampleRate /= stride;
  }
  // Judge silence on the original signal: flattening normalises loudness away,
  // so gating afterwards would let room noise through as a confident note.
  let mean = 0;
  for (const n of buffer) mean += n;
  mean /= buffer.length;
  let power = 0;
  for (const n of buffer) power += (n - mean) ** 2;
  if (Math.sqrt(power / buffer.length) < 0.008) return null;
  buffer = flattenEnvelope(buffer, sampleRate);
  const maxLag = Math.min(
    Math.floor(sampleRate / MIN_HZ),
    Math.floor(buffer.length / 2) - 1,
  );
  const minLag = Math.max(2, Math.floor(sampleRate / MAX_HZ));
  const size = buffer.length - maxLag;
  const diff = new Float64Array(maxLag + 1);
  diff[0] = 1;
  let running = 0;
  for (let lag = 1; lag <= maxLag; lag++) {
    let sum = 0;
    for (let j = 0; j < size; j++) {
      const d = buffer[j] - buffer[j + lag];
      sum += d * d;
    }
    running += sum;
    diff[lag] = running === 0 ? 1 : (sum * lag) / running;
  }
  let lag = minLag;
  for (; lag < maxLag; lag++)
    if (diff[lag] < THRESHOLD) {
      while (lag + 1 < maxLag && diff[lag + 1] < diff[lag]) lag++;
      break;
    }
  if (lag >= maxLag || diff[lag] > THRESHOLD) return null;
  const a = diff[lag - 1],
    b = diff[lag],
    c = diff[lag + 1],
    den = a - 2 * b + c;
  const refined = lag + (den === 0 ? 0 : (a - c) / (2 * den));
  const frequency = sampleRate / refined;
  return frequency >= MIN_HZ && frequency <= MAX_HZ
    ? { frequency, confidence: 1 - b }
    : null;
}

// Keep at least C2–C6 visible, and expand by octaves for lower/higher singing.
export function pitchChartBounds(notes: number[]): [number, number] {
  const valid = notes.filter(Number.isFinite);
  return [
    Math.max(
      23,
      Math.min(35, Math.floor(Math.min(36, ...valid) / 12) * 12 - 1),
    ),
    Math.min(97, Math.max(85, Math.ceil(Math.max(84, ...valid) / 12) * 12 + 1)),
  ];
}

File: ios/Packages/SingwellCore/Sources/SingwellCore/Pitch.swift
import Foundation

/// Pitch maths shared by the live view, warm-ups and the quest.
public enum Pitch {
    /// Note names use the sharp symbol so labels stay compact on the piano roll.
    public static let noteNames = ["C", "C♯", "D", "D♯", "E", "F", "F♯", "G", "G♯", "A", "A♯", "B"]

    public static func midiToFrequency(_ midi: Double) -> Double {
        440 * pow(2, (midi - 69) / 12)
    }

    public static func frequencyToMidi(_ frequency: Double) -> Double {
        69 + 12 * log2(frequency / 440)
    }

    /// "D♯4" for MIDI 63. Rounds to the nearest semitone first.
    public static func noteName(_ midi: Double) -> String {
        let n = Int(midi.rounded())
        let index = ((n % 12) + 12) % 12
        let octave = Int((Double(n) / 12).rounded(.down)) - 1
        return noteNames[index] + String(octave)
    }

    /// Letter without the octave, for compact key labels.
    public static func pitchClass(_ midi: Int) -> String {
        noteNames[((midi % 12) + 12) % 12]
    }

    public static func isBlackKey(_ midi: Int) -> Bool {
        [1, 3, 6, 8, 10].contains(((midi % 12) + 12) % 12)
    }

    public struct Detection: Equatable, Sendable {
        public var frequency: Double
        public var confidence: Double
        public init(frequency: Double, confidence: Double) {
            self.frequency = frequency
            self.confidence = confidence
        }
    }

    /// Lowest fundamental treated as a voice: just under A1. Below is rumble, not singing.
    public static let minHz = 54.0
    /// Highest fundamental treated as a voice, comfortably above any sung note.
    public static let maxHz = 2200.0

    /// Flatten a slow amplitude envelope while leaving the pitch periodicity alone.
    /// A lip trill flutters loudness at roughly 20–35 Hz; on low notes that flutter beats the
    /// vocal period and YIN reports a note two octaves down. Dividing by a short running
    /// magnitude equalises loudness so only the vocal period is left. Vibrato passes through.
    static func flattenEnvelope(_ buffer: [Float], sampleRate: Double) -> [Float] {
        let window = max(8, Int(sampleRate / 500))
        var magnitude = [Float](repeating: 0, count: buffer.count)
        var sum: Float = 0
        for i in buffer.indices {
            sum += abs(buffer[i])
            if i >= window { sum -= abs(buffer[i - window]) }
            magnitude[i] = sum / Float(min(i + 1, window))
        }
        let peak = magnitude.max() ?? 0
        let floor = peak * 0.15
        var out = [Float](repeating: 0, count: buffer.count)
        for i in buffer.indices { out[i] = buffer[i] / max(magnitude[i], floor) }
        return out
    }

    /// YIN with cumulative mean normalisation and parabolic interpolation.
    /// Returns nil for silence, noise, or anything outside the voice band.
    public static func detect(_ input: [Float], sampleRate inputRate: Double) -> Detection? {
        guard inputRate.isFinite, inputRate > 0, input.count >= 1024 else { return nil }
        var buffer = input
        var sampleRate = inputRate
        // Average blocks before decimation at high device sample rates.
        let stride = max(1, Int(sampleRate / 48000))
        if stride > 1 {
            let reducedCount = buffer.count / stride
            var reduced = [Float](repeating: 0, count: reducedCount)
            for i in 0..<reducedCount {
                var sum: Float = 0
                for j in 0..<stride { sum += buffer[i * stride + j] }
                reduced[i] = sum / Float(stride)
            }
            buffer = reduced
            sampleRate /= Double(stride)
        }
        let count = buffer.count
        var mean: Double = 0
        for v in buffer { mean += Double(v) }
        mean /= Double(count)
        var power: Double = 0
        for v in buffer { let d = Double(v) - mean; power += d * d }
        // Judge silence on the original signal: flattening normalises loudness away.
        if (power / Double(count)).squareRoot() < 0.008 { return nil }
        buffer = flattenEnvelope(buffer, sampleRate: sampleRate)

        let maxLag = min(Int(sampleRate / minHz), count / 2 - 1)
        let minLag = max(2, Int(sampleRate / maxHz))
        guard maxLag > minLag + 1 else { return nil }
        let size = count - maxLag
        var diff = [Double](repeating: 1, count: maxLag + 1)
        var running: Double = 0
        buffer.withUnsafeBufferPointer { p in
            for lag in 1...maxLag {
                var sum: Float = 0
                var j = 0
                while j < size {
                    let d = p[j] - p[j + lag]
                    sum += d * d
                    j += 1
                }
                running += Double(sum)
                diff[lag] = running == 0 ? 1 : (Double(sum) * Double(lag)) / running
            }
        }
        var lag = minLag
        while lag < maxLag {
            if diff[lag] < 0.15 {
                while lag + 1 < maxLag && diff[lag + 1] < diff[lag] { lag += 1 }
                break
            }
            lag += 1
        }
        if lag >= maxLag || diff[lag] > 0.15 { return nil }
        let a = diff[lag - 1], b = diff[lag], c = diff[lag + 1]
        let den = a - 2 * b + c
        let refined = Double(lag) + (den == 0 ? 0 : (a - c) / (2 * den))
        let frequency = sampleRate / refined
        guard frequency >= minHz, frequency <= maxHz else { return nil }
        return Detection(frequency: frequency, confidence: 1 - b)
    }

    /// Keep at least C2–C6 visible and expand by octaves for lower or higher singing.
    public static func chartBounds(_ notes: [Double]) -> (low: Int, high: Int) {
        let valid = notes.filter { $0.isFinite }
        let lowest = min(36, valid.min() ?? 36)
        let highest = max(84, valid.max() ?? 84)
        let low = max(23, min(35, Int((lowest / 12).rounded(.down)) * 12 - 1))
        let high = min(97, max(85, Int((highest / 12).rounded(.up)) * 12 + 1))
        return (low, high)
    }

    /// Signed cents from the nearest semitone, rounded, or nil when silent.
    public static func cents(_ midi: Double?) -> Int? {
        guard let midi else { return nil }
        return Int(((midi - midi.rounded()) * 100).rounded())
    }
}
